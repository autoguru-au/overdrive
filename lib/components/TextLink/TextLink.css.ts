import { style } from '@vanilla-extract/css';

import { themeContractVars as vars } from '../../themes/theme.css';

export const root = style({
	transitionDelay: '0s',
	transitionTimingFunction: 'cubic-bezier(0, 0, 0.2, 1)',
	transitionDuration: '0.2s',
	transitionProperty: 'box-shadow',
	boxShadow: `inset 0 0px 0 0 ${vars.typography.colour.link}`,
	':hover': {
		boxShadow: `inset 0 -2px 0 0 ${vars.typography.colour.link}`,
	},
});

export const icon = style({
	transition: `transform 0.2s ${vars.animation.easing.decelerate} 0s`,
	right: 0,
	top: '50%',
	transform: 'translateY(-50%)',
});

export const muted = style({
	':hover': {
		color: 'white',
		boxShadow: `inset 0 -1.6em 0 0 ${vars.typography.colour.link}`,
	},
});
