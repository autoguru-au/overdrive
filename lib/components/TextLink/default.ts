export { TextLink as default } from './TextLink';
