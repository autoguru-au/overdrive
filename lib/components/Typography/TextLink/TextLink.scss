@import '../../../theme/tokens/facade';

.root {
	box-shadow: inset 0 -2px 0 0 getColour(green, 400);
	transition: box-shadow 0.2s cubic-bezier(0, 0, 0.2, 1) 0s;

	&.strong {
		box-shadow: inset 0 -3px 0 0 getColour(green, 400);
	}

	&:hover {
		box-shadow: inset 0 -1.6em 0 0 getColour(green, 400);
	}
}

.anchor {
	outline: none;
	text-decoration: none;
}
