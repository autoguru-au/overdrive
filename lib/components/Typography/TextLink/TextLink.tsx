import { invariant } from '@autoguru/utilities';
import clsx from 'clsx';
import React, {
	AnchorHTMLAttributes,
	cloneElement,
	ComponentProps,
	forwardRef,
	ReactElement,
	ReactText,
	RefForwardingComponent,
} from 'react';

import { Text } from '..';
import styles from './TextLink.scss';

type TextProps = Omit<ComponentProps<typeof Text>, 'is' | 'white'>;
type AnchorProps = Omit<
	AnchorHTMLAttributes<HTMLAnchorElement>,
	'children' | 'style' | 'as' | keyof TextProps
>;

interface Props extends TextProps, AnchorProps {
	children: ReactText;
	className?: string;
	as?: ReactElement;
}

export const TextLink: RefForwardingComponent<
	HTMLAnchorElement,
	Props
> = forwardRef(
	(
		{ as, children, className, strong = true, muted, size, ...props },
		ref,
	) => {
		invariant(
			as !== void 0 && props.href !== void 0,
			'You cannot have both href and as defined.',
		);

		const body = (
			<Text
				is="span"
				muted={muted}
				size={size}
				strong={strong}
				className={clsx(styles.root, {
					[styles.strong]: strong,
				})}>
				{children}
			</Text>
		);

		if (as === void 0) {
			return (
				<a
					ref={ref as any}
					rel={props.rel || 'noopener noreferrer'}
					className={className}
					{...props}>
					{body}
				</a>
			);
		}

		return cloneElement(as, { ref, className }, body);
	},
);
