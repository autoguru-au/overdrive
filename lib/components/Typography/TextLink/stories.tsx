import { boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import { Text } from '..';
import { TextLink } from '.';

storiesOf('Foundation|Typography/TextLink', module)
	.addDecorator(story => (
		<div style={{ width: '100%', maxWidth: 300 }}>{story()}</div>
	))
	.add('Standard', () => (
		<>
			<Text muted is="p">
				Lorem ipsum dolor sit amet, consectetur adipisicing elit. Ad,{' '}
				<TextLink muted href="test" strong={false}>
					Hello
				</TextLink>{' '}
				autem consectetur consequuntur eius ex{' '}
				<TextLink href="test" strong={boolean('Strong', true)}>
					Hello
				</TextLink>{' '}
				fugiat illo ipsum nobis numquam, officiis placeat quia, quidem
				reprehenderit rerum temporibus veniam vero.
			</Text>
		</>
	));
