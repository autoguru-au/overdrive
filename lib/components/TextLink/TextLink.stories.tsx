import { ArrowRightIcon, ChevronRightIcon } from '@autoguru/icons';
import { Meta, StoryObj } from '@storybook/react';
import React, { type ComponentProps } from 'react';

import { Heading } from '../Heading/Heading';
import { Text } from '../Text/Text';

import { TextLink } from './TextLink';

const sizeScale = ['1', '2', '3', '4', '5', '6', '7', '8', '9'];
const fontWeightOptions = ['normal', 'semiBold', 'bold'];

const noWrapOptions: Array<ComponentProps<typeof Heading>['noWrap']> = [
	false,
	true,
];
const transformOptions: Array<ComponentProps<typeof Text>['transform']> = [
	'uppercase',
	'capitalize',
	undefined,
];

const meta = {
	title: 'Primatives/Text Link',
	component: TextLink,
	decorators: [
		(Story) => (
			<div style={{ maxWidth: 300 }}>
				<Story />
			</div>
		),
	],
	args: {
		size: '4',
		fontWeight: 'semiBold',
		icon: undefined,
		muted: false,
		noWrap: undefined,
		transform: undefined,
		href: '#link',
		children: 'Hello',
	},
	argTypes: {
		icon: {
			defaultValue: null,
			description: 'Input field Icon',
			options: ['Arrow Right', 'Chevron Right'],
			mapping: {
				'Arrow Right': ArrowRightIcon,
				'Chevron Right': ChevronRightIcon,
			},
		},
		noWrap: {
			options: noWrapOptions,
			defaultValue: false,
			control: {
				type: 'boolean',
			},
		},
		transform: {
			options: transformOptions,
			defaultValue: null,
			control: {
				type: 'select',
			},
		},
		fontWeight: {
			options: fontWeightOptions,
			defaultValue: null,
			control: {
				type: 'select',
			},
		},
		size: {
			options: sizeScale,
			defaultValue: void 0,
			control: {
				type: 'select',
			},
		},
	},
} satisfies Meta<typeof TextLink>;

export default meta;

type Story = StoryObj<typeof TextLink>;

export const Standard: Story = {};

export const InsideParagraph: Story = {
	decorators: [
		(Story) => (
			<Text as="p">
				Lorem ipsum dolor sit amet, consectetur adipisicing elit. Ad,{' '}
				<Story /> autem consectetur consequuntur eius fugiat illo ipsum
				nobis numquam, officiis placeat quia, quidem reprehenderit rerum
				temporibus veniam vero.
			</Text>
		),
	],
};

export const WithIcon: Story = {
	args: {
		icon: ChevronRightIcon,
	},
};

export const WithIconInsideParagraph: Story = {
	args: {
		icon: ArrowRightIcon,
	},
	decorators: InsideParagraph.decorators,
};
