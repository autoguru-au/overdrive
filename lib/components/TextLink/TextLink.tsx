import { IconType } from '@autoguru/icons';
import { invariant } from '@autoguru/utilities';
import clsx from 'clsx';
import * as React from 'react';
import {
	AnchorHTMLAttributes,
	cloneElement,
	ComponentProps,
	createElement,
	ElementType,
	forwardRef,
	isValidElement,
	ReactElement,
	ReactNode,
} from 'react';

import { Box } from '../Box/Box';
import { boxStyles } from '../Box/boxStyles';
import { Icon } from '../Icon/Icon';
import { Text } from '../Text/Text';

import * as styles from './TextLink.css';

type TextProps = Omit<ComponentProps<typeof Text>, 'is' | 'colour'>;
type AnchorProps = Omit<
	AnchorHTMLAttributes<HTMLAnchorElement>,
	'children' | 'color' | 'style' | 'is' | keyof TextProps
>;

export interface TextLinkProps extends TextProps, AnchorProps {
	children?: ReactNode;
	className?: string;
	is?: ElementType | ReactElement;
	muted?: boolean;
	icon?: IconType;
}

export const TextLink = forwardRef<HTMLAnchorElement, TextLinkProps>(
	(
		{
			is: Component,
			children,
			className,
			strong,
			fontWeight = 'semiBold',
			muted = false,
			size,
			icon,
			...props
		},
		ref,
	) => {
		invariant(
			!(Component !== undefined && props.href !== undefined),
			'You cannot have both href and as defined.',
		);
		const body = (
			<Text
				as="span"
				colour={muted ? 'muted' : 'link'}
				size={size}
				fontWeight={fontWeight}
				strong={strong}
				className={clsx(
					boxStyles({
						as: 'span',
						pointerEvents: 'none',
						position: 'relative',
						paddingRight: icon ? '5' : void 0,
					}),
					{
						[styles.muted]: muted,
					},
				)}
			>
				{children}
				{icon ? (
					<Icon
						icon={icon}
						size="small"
						display="inline-block"
						className={clsx(
							styles.icon,
							boxStyles({
								position: 'absolute',
							}),
						)}
					/>
				) : null}
			</Text>
		);

		const allProps = {
			rel: props.rel ?? 'noopener noreferrer',
			...props,
			ref,
		};

		if (Component === undefined) {
			return (
				<Box as="a" className={[className, styles.root]} {...allProps}>
					{body}
				</Box>
			);
		}

		return isValidElement(Component)
			? cloneElement(Component, allProps, body)
			: createElement(Component, allProps, body);
	},
);

TextLink.displayName = 'TextLink';
