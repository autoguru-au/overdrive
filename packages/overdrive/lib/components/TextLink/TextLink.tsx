import { invariant } from '@autoguru/utilities';
import clsx from 'clsx';
import * as React from 'react';
import {
	AnchorHTMLAttributes,
	cloneElement,
	ComponentProps,
	createElement,
	ElementType,
	forwardRef,
	isValidElement,
	ReactElement,
	ReactText,
} from 'react';
import { useStyles } from 'react-treat';

import { Box, useBoxStyles } from '../Box';
import { Text } from '../Text';
import * as styleRefs from './TextLink.treat';

type TextProps = Omit<ComponentProps<typeof Text>, 'is' | 'colour'>;
type AnchorProps = Omit<
	AnchorHTMLAttributes<HTMLAnchorElement>,
	'children' | 'style' | 'is' | keyof TextProps
>;

export interface Props extends TextProps, AnchorProps {
	children: ReactText;
	className?: string;
	is?: ElementType | ReactElement;
	muted?: boolean;
}

export const TextLink = forwardRef<HTMLAnchorElement, Props>(
	(
		{
			is: Component,
			children,
			className,
			fontWeight = 'bold',
			strong,
			muted = false,
			size,
			...props
		},
		ref,
	) => {
		invariant(
			!(Component !== undefined && props.href !== undefined),
			'You cannot have both href and as defined.',
		);

		const styles = useStyles(styleRefs);

		const body = (
			<Text
				is="span"
				colour={muted && ('muted' as any)}
				size={size}
				strong={strong}
				fontWeight={fontWeight}
				className={clsx(
					styles.root,
					useBoxStyles({ pointerEvents: 'none' }),
					{
						[styles.muted]: muted,
					},
				)}>
				{children}
			</Text>
		);

		const allProps = {
			rel: props.rel ?? 'noopener noreferrer',
			...props,
			ref,
			className: clsx(className),
		};

		if (Component === undefined) {
			return (
				<Box is="a" {...allProps}>
					{body}
				</Box>
			);
		}

		return isValidElement(Component)
			? cloneElement(Component, allProps, body)
			: createElement(Component, allProps, body);
	},
);
