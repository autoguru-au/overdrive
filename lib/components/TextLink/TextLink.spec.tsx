import { render, screen } from '@testing-library/react';
import React from 'react';
import { describe, expect, it } from 'vitest';

import { TextLink } from './TextLink';

describe('<TextLink />', () => {
	it('should render as anchor by default', () => {
		render(<TextLink href="/test">Link text</TextLink>);

		const link = screen.getByRole('link');
		expect(link.tagName).toBe('A');
		expect(link).toHaveAttribute('href', '/test');
		expect(link).toHaveTextContent('Link text');
	});

	it('should render with different element using as prop', () => {
		render(
			<TextLink as="button" data-testid="link-button">
				Button link
			</TextLink>,
		);

		const button = screen.getByTestId('link-button');
		expect(button.tagName).toBe('BUTTON');
		expect(button).toHaveTextContent('Button link');
	});
});
