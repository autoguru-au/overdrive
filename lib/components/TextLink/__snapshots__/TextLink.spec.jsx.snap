// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TextLink /> > should allow as override 1`] = `
<div
  class="theme_OD_Base__j49br80"
  data-od-component-provider=""
>
  <p
    class="reset_resetVariants__4z7xt3c reset_resetVariants_as_p__4z7xt3e reset_trimmedElement__4z7xt30 TextLink_root__5ax1bk0"
    data-od-component="textlink"
    rel="nooperner noreferrer"
  >
    <span
      class="reset_resetVariants__4z7xt3c reset_resetVariants_as_span__4z7xt3l  sprinkles_pointerEvents_none__onxwju2u sprinkles_position_relative_mobile__onxwju8v reset_resetVariants__4z7xt3c reset_resetVariants_as_span__4z7xt3l  sprinkles_fontSize_4__onxwju21 sprinkles_lineHeight_4__onxwju2a sprinkles_fontWeight_semiBold__onxwju2h sprinkles_color_link__onxwju6k typography_typographyStyles__yhjp330"
    >
      Test
    </span>
  </p>
</div>
`;

exports[`<TextLink /> > should match the snapshot 1`] = `
<div
  class="theme_OD_Base__j49br80"
  data-od-component-provider=""
>
  <a
    class="reset_resetVariants__4z7xt3c reset_resetVariants_as_a__4z7xt316 reset__4z7xt32 focusOutline_focusOutlineStyle__tq83zq0 TextLink_root__5ax1bk0"
    data-od-component="textlink"
    rel="nooperner noreferrer"
  >
    <span
      class="reset_resetVariants__4z7xt3c reset_resetVariants_as_span__4z7xt3l  sprinkles_pointerEvents_none__onxwju2u sprinkles_position_relative_mobile__onxwju8v reset_resetVariants__4z7xt3c reset_resetVariants_as_span__4z7xt3l  sprinkles_fontSize_4__onxwju21 sprinkles_lineHeight_4__onxwju2a sprinkles_fontWeight_semiBold__onxwju2h sprinkles_color_link__onxwju6k typography_typographyStyles__yhjp330"
    />
  </a>
</div>
`;
